File: packages/quran-engine-swift/Sources/QuranEngine/WordByWord.swift
import Foundation

/// Word by word: what each word of an ayah means, and how it is said.
///
/// Two layers over the SAME tokens - the English gloss and a Latin transliteration - where the
/// tokens are the ayah's own whitespace-separated words. Split the ayah and index straight in; the
/// alignment against a corpus that tokenizes ~200 ayahs differently was done once, at build time.
///
/// A token with no word of its own (the ۞ ornament, the tail of a word the corpus writes as two)
/// carries "" in both layers - show nothing for it rather than a neighbour's meaning.
///
/// See docs/12-word-by-word.md.
public struct WordByWordPack: Decodable, Sendable {
    /// Surah id -> ayahs in id order -> one entry per token.
    public let english: [String: [[String]]]
    public let transliteration: [String: [[String]]]
}

public struct GlossedWord: Sendable, Equatable {
    /// 1-based index of the word in the ayah.
    public let position: Int
    /// The ayah's own token.
    public let arabic: String
    /// The gloss, "" when the token has none.
    public let english: String
    public let transliteration: String
}

public final class WordByWord {
    public enum Layer { case english, transliteration }

    private let englishLayer: [String: [[String]]]
    private let latinLayer: [String: [[String]]]
    private weak var quran: Quran?

    public init(pack: WordByWordPack? = nil, quran: Quran? = nil) {
        self.englishLayer = pack?.english ?? [:]
        self.latinLayer = pack?.transliteration ?? [:]
        self.quran = quran
    }

    /// Whether a pack is loaded at all - cheap enough to gate UI on.
    public var isLoaded: Bool { !englishLayer.isEmpty }

    /// Every word of an ayah, in reading order.
    public func words(_ surahId: Int, _ ayahId: Int) -> [GlossedWord] {
        guard let english = glosses(surahId, ayahId) else { return [] }
        let latin = transliterations(surahId, ayahId) ?? []
        let tokens = self.tokens(surahId, ayahId)
        return english.enumerated().map { index, gloss in
            GlossedWord(position: index + 1,
                        arabic: index < tokens.count ? tokens[index] : "",
                        english: gloss,
                        transliteration: index < latin.count ? latin[index] : "")
        }
    }

    /// One word, by its 1-based position.
    public func word(_ surahId: Int, _ ayahId: Int, position: Int) -> GlossedWord? {
        let all = words(surahId, ayahId)
        return position >= 1 && position <= all.count ? all[position - 1] : nil
    }

    public func glosses(_ surahId: Int, _ ayahId: Int) -> [String]? {
        row(englishLayer, surahId, ayahId)
    }

    public func transliterations(_ surahId: Int, _ ayahId: Int) -> [String]? {
        row(latinLayer, surahId, ayahId)
    }

    public func entries(_ layer: Layer, _ surahId: Int, _ ayahId: Int) -> [String]? {
        layer == .english ? glosses(surahId, ayahId) : transliterations(surahId, ayahId)
    }

    /// Ayahs containing a word whose gloss carries `term` - a word-level English search, which finds
    /// ayahs a translation search misses because no translator used that phrasing.
    public func find(_ term: String, limit: Int = 50) -> [(surah: Int, ayah: Int, position: Int, english: String, transliteration: String)] {
        let needle = term.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !needle.isEmpty else { return [] }
        var out: [(surah: Int, ayah: Int, position: Int, english: String, transliteration: String)] = []
        for surahKey in englishLayer.keys.compactMap(Int.init).sorted() {
            let rows = englishLayer["\(surahKey)"] ?? []
            for (ayahIndex, glosses) in rows.enumerated() {
                for (index, gloss) in glosses.enumerated() where gloss.lowercased().contains(needle) {
                    let latinRows = latinLayer["\(surahKey)"] ?? []
                    let latin = ayahIndex < latinRows.count && index < latinRows[ayahIndex].count
                        ? latinRows[ayahIndex][index] : ""
                    out.append((surah: surahKey, ayah: ayahIndex + 1, position: index + 1,
                                english: gloss, transliteration: latin))
                    if out.count >= limit { return out }
                }
            }
        }
        return out
    }

    private func row(_ layer: [String: [[String]]], _ surahId: Int, _ ayahId: Int) -> [String]? {
        guard let rows = layer["\(surahId)"], ayahId >= 1, ayahId <= rows.count else { return nil }
        return rows[ayahId - 1]
    }

    private func tokens(_ surahId: Int, _ ayahId: Int) -> [String] {
        guard let text = quran?.ayah(surahId, ayahId)?.textArabic else { return [] }
        return text.split(whereSeparator: { $0.isWhitespace }).map(String.init)
    }
}

File: packages/quran-engine-py/quran_engine/word_by_word.py
"""Word by word: what each word means, and how it is said. Mirrors src/wordByWord.js.

Two layers over the SAME tokens - the English gloss and a Latin transliteration - where the tokens
are the ayah's own whitespace-separated words. Split the ayah and index straight in; the alignment
against a corpus that tokenizes ~200 ayahs differently was done once, at build time.

A token with no word of its own (the ۞ ornament, the tail of a merged word) carries "" in both.
"""
from __future__ import annotations
import re
from typing import Optional

_WHITESPACE = re.compile(r"\s+")


class Word(dict):
    """position, arabic, english, transliteration."""


class WordByWord:
    def __init__(self, english: Optional[dict] = None, transliteration: Optional[dict] = None,
                 quran=None):
        self._english = english or {}
        self._transliteration = transliteration or {}
        self._quran = quran

    @property
    def is_loaded(self) -> bool:
        return bool(self._english)

    def words(self, surah_id: int, ayah_id: int) -> list[Word]:
        english = self.glosses(surah_id, ayah_id)
        if english is None:
            return []
        latin = self.transliterations(surah_id, ayah_id) or []
        tokens = self._tokens(surah_id, ayah_id)
        return [Word({
            "position": i + 1,
            "arabic": tokens[i] if i < len(tokens) else "",
            "english": gloss,
            "transliteration": latin[i] if i < len(latin) else "",
        }) for i, gloss in enumerate(english)]

    def word(self, surah_id: int, ayah_id: int, position: int) -> Optional[Word]:
        words = self.words(surah_id, ayah_id)
        return words[position - 1] if 1 <= position <= len(words) else None

    def glosses(self, surah_id: int, ayah_id: int) -> Optional[list[str]]:
        return self._row(self._english, surah_id, ayah_id)

    def transliterations(self, surah_id: int, ayah_id: int) -> Optional[list[str]]:
        return self._row(self._transliteration, surah_id, ayah_id)

    def find(self, term: str, limit: int = 50) -> list[dict]:
        """Ayahs with a word whose gloss carries `term` - a word-level English search."""
        needle = term.strip().lower()
        if not needle:
            return []
        out: list[dict] = []
        for surah in sorted(self._english, key=int):
            for ayah_index, glosses in enumerate(self._english[surah]):
                for i, gloss in enumerate(glosses):
                    if needle not in gloss.lower():
                        continue
                    rows = self._transliteration.get(surah) or []
                    latin = rows[ayah_index][i] if ayah_index < len(rows) and i < len(rows[ayah_index]) else ""
                    out.append({"surah": int(surah), "ayah": ayah_index + 1, "position": i + 1,
                                "english": gloss, "transliteration": latin})
                    if len(out) >= limit:
                        return out
        return out

    @staticmethod
    def _row(layer: dict, surah_id: int, ayah_id: int) -> Optional[list[str]]:
        rows = layer.get(str(surah_id))
        if not rows or ayah_id < 1 or ayah_id > len(rows):
            return None
        return rows[ayah_id - 1]

    def _tokens(self, surah_id: int, ayah_id: int) -> list[str]:
        if self._quran is None:
            return []
        ayah = self._quran.ayah(surah_id, ayah_id)
        return _WHITESPACE.split(ayah.text_arabic.strip()) if ayah else []
